algorithms/rainbow/agent.py

```diff
@@ -0,0 +1,158 @@
+        Z is represented with a fixed number of "atoms", which are discrete positions equidistant
+        along its support defined between Vmin-Vmax.
+
+        As an example, for a given (s,a) pair, we can represent Z(s,a) with 8 atoms as follows:
+
+                   .        .     .
+                .  |     .  |  .  |
+                |  |  .  |  |  |  |  .
+                |  |  |  |  |  |  |  |
+           Vmin ----------------------- Vmax
+        """
+        self.action_space = env.action_space
+        self.atoms = args.atoms
+        self.Vmin = args.V_min
+        self.Vmax = args.V_max
+        self.support = torch.linspace(args.V_min, args.V_max, self.atoms).to(device=args.device)
+        self.delta_z = (args.V_max - args.V_min) / (self.atoms - 1)
+        self.batch_size = args.batch_size
+        self.n = args.multi_step
+        self.discount = args.discount
+
+        self.online_net = RainbowDQN(args, self.action_space).to(device=args.device)
+        if args.model and os.path.isfile(args.model):
+            """
+            Always load tensors onto CPU by default, will shift to GPU if necessary to avoid 
+            GPU RAM surge when loading a model checkpoint as recommended in pytorch official 
+            documentation.
+            Source: https://pytorch.org/docs/stable/torch.html#torch.load
+            """
+            self.online_net.load_state_dict(torch.load(args.model, map_location='cpu'))
+        self.online_net.train()
+
+        self.target_net = RainbowDQN(args, self.action_space).to(device=args.device)
+        self.update_target_net()
+        self.target_net.train()
+        for param in self.target_net.parameters():
+            param.requires_grad = False
+
+        self.optimiser = optim.Adam(self.online_net.parameters(), lr=args.lr, eps=args.adam_eps)
+
+    def reset_noise(self):
+        """ Resets noisy weights in all linear layers (of online net only) """
+        self.online_net.reset_noise()
+
+    def act(self, state):
+        """ Acts based on single state (no batch) """
+        with torch.no_grad():
+            return (self.online_net(state.unsqueeze(0)) * self.support).sum(2).argmax(1).item()
+
+    def act_e_greedy(self, state, epsilon=0.001):
+        """
+        Acts with an ε-greedy policy (used for evaluation only)
+        High ε can reduce evaluation scores drastically
+        """
+        return np.random.randint(0, self.action_space.n) if np.random.random() < epsilon \
+            else self.act(state)
+
+    def learn(self, mem):
+        """
+        Executes 1 gradient descent step sampling batch_size transitions from the memory
+        """
+        # Sample transitions
+        idxs, states, actions, returns, next_states, nonterminals, weights = \
+          mem.sample(self.batch_size)
+
+        # Calculate current state probabilities (online network noise already sampled)
+        log_ps = self.online_net(states, log=True)  # Log probabilities log p(s_t, ·; θonline)
```

Comment:
log_ps? ps? Are not returning Q?

From the blog, they mention the net returning |A| x num_atoms matrix.

Reply:
I added this comment on model.py `log softmax used while learning to generate probabilities with higher numerical stability`. Let me know if that is enough for you.

Reply:
why do we need higher numerical stability in training only? Because training can get crazy and blow up things?
Why would you ever choose the non-log softmax version?

Reply:
This is supporting the moving target DQN. So you are chasing an optimum that is moving and you want to add stability. During evaluation you don't care anymore because that is not going to be "propagated knowledge". Ideally I think you don't want the log, but it helps like an extra learning rate (kind of, please take it as an informal statement and not taken from a calculus book).

Reply:
Ok I accept your informality. 

maybe another reason is that we only need logs for backprop and many other mathematical formulations, not really for just choosing the argmax. Maybe add that in as well? e.g. `log is not needed for evaluation `